nuxeo-retention-web/elements/nuxeo-retention-analytics.js

```diff
@@ -0,0 +1,351 @@
+          ecm-mixin-type="HasRelatedText"
+          where='{"ecm:hasLegalHold": "true"}'
+          metrics="cardinality(ecm:uuid)"
+          data="{{legalHoldCount}}"
+          index="[[index]]"
+        >
+        </nuxeo-repository-data>
+
+        <nuxeo-card class="infoCard">
+          <div class="rangeSelector">
+            <span class="mainRetentionInfo"
+              >[[i18n('retentionAnalytics.underActiveRetention')]]
+              <span class="retentionCount">[[activeRetentionCount]]</span>
+            </span>
+            <span class="mainRetentionInfo"> | </span>
+            <span class="mainRetentionInfo"
+              >[[i18n('retentionAnalytics.legalHoldCount')]]
+              <span class="retentionCount">[[legalHoldCount]]</span>
+            </span>
+          </div>
+          <div class="rangeSelector">
+            <div class="dateLabel">
+              <label>[[i18n('retentionAnalytics.from')]]</label>
+            </div>
+            <nuxeo-date-picker value="{{startDate}}"></nuxeo-date-picker>
+            <div class="dateLabel">
+              <label>[[i18n('retentionAnalytics.to')]]</label>
+            </div>
+            <nuxeo-date-picker value="{{endDate}}"></nuxeo-date-picker>
+          </div>
+        </nuxeo-card>
+
+        <div class="flex-layout">
+          <!-- Retention by Document Type -->
+          <nuxeo-repository-data
+            start-date="[[_formatDate(startDate)]]"
+            end-date="[[_extendEndDate(endDate)]]"
+            date-field="dc:created"
+            grouped-by="ecm:primaryType"
+            where='{"ecm:isRecord": "true"}'
+            group-limit="20"
+            data="{{allDocCategoriesCount}}"
+            index="[[index]]"
+          >
+          </nuxeo-repository-data>
+
+          <nuxeo-card heading="[[i18n('retentionAnalytics.retentionByDocType')]]">
+            <chart-pie
+              values="[[_values(allDocCategoriesCount)]]"
+              labels="[[_labels(allDocCategoriesCount)]]"
+              options='{ "legend": { "display": true, "position": "bottom", "labels": { "boxWidth": 12 } }, "animation": false }'
+            >
+            </chart-pie>
+          </nuxeo-card>
+
+          <!-- RETENTION STATUS -->
+          <!-- No Retention -->
+          <nuxeo-repository-data
+            start-date="[[_formatDate(startDate)]]"
+            end-date="[[_extendEndDate(endDate)]]"
+            date-field="dc:created"
+            ecm-mixin-type="HasRelatedText"
+            metrics="cardinality(ecm:uuid)"
+            where='{"ecm:isRecord": "false"}'
+            data="{{noRetention}}"
+            index="[[index]]"
+          >
+          </nuxeo-repository-data>
+
+          <!-- Future Retentions -->
+          <nuxeo-repository-data
+            start-date="[[_formatDate(startDate)]]"
+            end-date="[[_extendEndDate(endDate)]]"
+            date-field="dc:created"
+            grouped-by="ecm:retainUntil"
```

Comment:
why you group by `"ecm:retainUntil"` which is a datetime. The Future rentention are the ones with `ecm:retainUntil!=null` and in the future. You have two ways to do it:

- Reuse the `legalHoldCount` which compute the number of documents under legal hod.
- If we want only those with  a retain date in the future, You can try things:

```
 <nuxeo-repository-data
            start-date="[[_formatDate(startDate)]]"
            end-date="[[_extendEndDate(endDate)]]"
            date-field="dc:created"
            where='[[filter]]'
            metrics="cardinality(ecm:uuid)"
            group-limit="10"
            data="{{futureRetention}}"
            index="[[index]]"
          >
          </nuxeo-repository-data>
```

where filter is:

```
 static get properties() {
    return {
      filter: {
        type: Array,
        value: [{"range": {"ecm:retainUntil": {"gte": moment().format('YYYY-MM-DD')}}}]
      },
```

and we will only compute the number of the documents: `metrics="cardinality(ecm:uuid)"`

Note that, you have to update your method:

```
  _valuesRetentionStatus(noRetention, futureRetention, expiredRetention) {
    //const interminate = futureRetention.filter((entry) => entry.key === '9999-01-01T00:00:00.000Z').length;
    return [[noRetention, futureRetention, expiredRetention.length, 10]];
  }
```


Now for the  indeterminate ones, it seems that when the date is `'9999-01-01T00:00:00.000+00:00'` it means the document is  indeterminate. Perhaps doing two queries is more performant than requested the server and then iterate over them and filter again. Doing two queries can be a good approche where we compute directly the metrics `cardinality`

Reply:
good idea, i'll try that.

Reply:
done